pytorch_lightning/callbacks/prediction_writer.py

```diff
@@ -0,0 +1,118 @@
+
+
+class BasePredictionWriter(Callback):
+    """
+    Base class to implement how the predictions should be stored.
+
+    Args:
+        write_interval: When to write.
+
+    Example::
+
+        from pytorch_lightning.callbacks import BasePredictionWriter
+
+        class CustomWriter(BasePredictionWriter):
+
+            def __init__(self, output_dir: str, write_interval: str):
+                super().__init__(write_interval)
+                self.output_dir
+
+            def write_on_batch_end(
+                self, trainer, pl_module: 'LightningModule', prediction: Any, batch_indices: List[int], batch: Any,
+                batch_idx: int, dataloader_idx: int
+            ):
+                torch.save(prediction, os.path.join(self.output_dir, dataloader_idx, f"{batch_idx}.pt")
+
+            def write_on_epoch_end(
+                self, trainer, pl_module: 'LightningModule', predictions: List[Any], batch_indices: List[Any]
+            ):
+                torch.save(predictions, os.path.join(self.output_dir, "predictions.pt")
+    """
+
+    def __init__(self, write_interval: str = "batch") -> None:
+        if write_interval not in list(WriteInterval):
+            raise MisconfigurationException(f"`write_interval` should be one of {[i.value for i in WriteInterval]}.")
+        self.interval = WriteInterval(write_interval)
+
+    @abstractmethod
+    def write_on_batch_end(
+        self,
+        trainer: 'pl.Trainer',
+        pl_module: 'pl.LightningModule',
+        prediction: Any,
+        batch_indices: Optional[List[int]],
+        batch: Any,
+        batch_idx: int,
+        dataloader_idx: int,
+    ) -> None:
+        """Override with logic to write a single batch."""
+
+    @abstractmethod
+    def write_on_epoch_end(
+        self,
+        trainer: 'pl.Trainer',
+        pl_module: 'pl.LightningModule',
+        predictions: List[Any],
+        batch_indices: Optional[List[Any]],
+    ) -> None:
+        """Override with logic to write all batches."""
+
+    def on_predict_batch_end(
+        self,
+        trainer: 'pl.Trainer',
+        pl_module: 'pl.LightningModule',
+        outputs: Any,
+        batch: Any,
+        batch_idx: int,
+        dataloader_idx: int,
+    ) -> None:
+        if not self.interval.on_batch:
+            return
+        batch_indices = trainer.predict_loop.batch_indices if trainer.accelerator_connector.is_distributed else None
+        self.write_on_batch_end(trainer, pl_module, outputs, batch_indices, batch, batch_idx, dataloader_idx)
+
+    def on_predict_epoch_end(self, trainer: 'pl.Trainer', pl_module: 'pl.LightningModule', outputs: List[Any]) -> None:
+        if not self.interval.on_epoch:
+            return
+        batch_indices = trainer.predict_loop.batch_indices if trainer.accelerator_connector.is_distributed else None
+        self.write_on_epoch_end(trainer, pl_module, trainer.predict_loop._predictions, batch_indices)
```

Comment:
`_predictions` is private to the predict loop but it's being accessed externally here. should predictions be part of the predict loop public API?

Reply:
Done !